example-apps/token_curated_registry/app/app.go

```diff
@@ -0,0 +1,164 @@
+package app
+
+import (
+	"encoding/json"
+	handle "github.com/cosmos/cosmos-academy/example-apps/token_curated_registry/auth"
+	dbl "github.com/cosmos/cosmos-academy/example-apps/token_curated_registry/db"
+	"github.com/cosmos/cosmos-academy/example-apps/token_curated_registry/types"
+	bam "github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/wire"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	"github.com/cosmos/cosmos-sdk/x/bank"
+	abci "github.com/tendermint/abci/types"
+	"github.com/tendermint/go-amino"
+	"github.com/tendermint/go-crypto"
+	cmn "github.com/tendermint/tmlibs/common"
+	dbm "github.com/tendermint/tmlibs/db"
+	"github.com/tendermint/tmlibs/log"
+)
+
+const (
+	appName = "Registry"
+)
+
+// Extended ABCI application
+type RegistryApp struct {
+	*bam.BaseApp
+
+	cdc *amino.Codec
+
+	minDeposit int64
+
+	applyStage int64
+
+	commitStage int64
+
+	revealStage int64
+
+	dispensationPct float64
+
+	quorum float64
+
+	// keys to access the substores
+	capKeyMain     *sdk.KVStoreKey
+	capKeyAccount  *sdk.KVStoreKey
+	capKeyListings *sdk.KVStoreKey
```

Comment:
This is exactly the type of functionality I was looking for. Thanks for bringing this up.

How efficient are these range queries?

I want to include these in EndBlocker so that I can remove the Apply and ClaimReward msgs and just handle that automatically, by making the range query and updating stores appropriately.

Reply:
Better than the sum of individual queries for sure, but not really sure. In any case you can put a limit on the number of iteration per individual endblocker and spread the operation over multiple blocks if needed